### PR TITLE
stream: add exception policy for midstream flows - v2

### PR DESCRIPTION
**Pull request by jufajardini**
This allows to set a midstream-policy that can:
- fail closed (stream.midstream-policy=drop-flow)
- fail open (stream.midstream-policy=pass-flow)
- bypass stream (stream.midstream-policy=bypass)
- do nothing (default behavior)

Usage and behavior:
If `stream.midstream-policy` is set then if Suricata identifies a midstream flow
it will apply the corresponding action associated with the policy.

No setting means Suricata will not apply such policies, either inspecting the 
flow (if `stream.midstream=true`) or ignoring it `stream.midstream=false`.

Task #5468

Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/5468

Previous PR:
https://github.com/OISF/suricata/pull/7727

Describe changes:
- add a stream.midstream-policy option that allows for controlling if a flow should be fully dropped, passed or bypassed. It is also possible to just ignore the policy (keeps normal behavior) if it is a midstream flow (in IPS mode)

Updates from last PR:
- add exception policies options to the config file
- make so that midstream policies are applied regardless of the `stream.midstream` setting
- add a `PKT_DROP_REASON_STREAM_MIDSTREAM` option for packet drop reasons

suricata-verify-pr: 916
https://github.com/OISF/suricata-verify/pull/916

### CODE REVIEW

File: src/stream-tcp.c [hunk 1]
```diff
@@ -925,7 +926,11 @@ static int StreamTcpPacketStateNone(ThreadVars *tv, Packet *p,
     /* SYN/ACK */
     } else if ((p->tcph->th_flags & (TH_SYN|TH_ACK)) == (TH_SYN|TH_ACK)) {
         if (!stream_config.midstream && stream_config.async_oneside == FALSE)
-            return 0;
+            if (stream_config.midstream_policy == EXCEPTION_POLICY_IGNORE)
```
Comment (victorjulien) [inline src/stream-tcp.c:929]: does this mean that `stream.midstream-policy=pass-flow` effectively also adds `stream.midstream=true`?

File: suricata.yaml.in [hunk 2]
```diff
@@ -1281,6 +1287,9 @@ defrag:
 
 flow:
   memcap: 128mb
+# memcap-policy: ignore       # Can be "drop-flow", "pass-flow", "bypass",
```
Comment (victorjulien) [inline suricata.yaml.in:1290]: I think these comment blocks inside the actual settings are a bit of an anti-pattern. With stream we have a block above it.

**Review by victorjulien — CHANGES_REQUESTED** [review]
Wonder if the CI failures are related to the more aggressive behavior in the flow-drop case.

File: src/stream-tcp.c (hunk 1 above)
Comment (jufajardini) [reply src/stream-tcp.c:929]: Indeed, with `stream.midstream-policy=pass-flow` it acts as if `stream.midstrem=true`, allowing for the engine to still log the flow etc, but without detection. That was what I understood that should happen 😅

**Comment by jufajardini** [comment]
> Wonder if the CI failures are related to the more aggressive behavior in the flow-drop case.

gonna investigate now 👀

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | test | baseline | % |
|--|--|--|--| 
| | tlpw1_stats_chk | | |
| .tcp.rst | 137182 | 105279 | 130.3% |
| | generic_stats_chk | | |
| .flow.end.tcp_state.syn_recv | 66 | 52 | 126.92% |
| .tcp.segment_memcap_drop | 33004 | 11729 | 281.39% |
| .tcp.insert_data_normal_fail | 32564 | 11358 | 286.71% |
| .app_layer.error.http.parser | 28 | 55 | 50.91% | 

Pipeline 8706
WARNING: THERE IS A KNOWN BAD BASELINE WITH PACKET DROPS. bE MINDFUL OF ANY RESULTS.

**Comment by jufajardini** [comment]
Replaced by: https://github.com/OISF/suricata/pull/7781

**Pull request closed without merging**